File: src/codecs/text/tcf-cli.ts
#!/usr/bin/env node

/**
 * TCF (Text Codec Format) CLI Tool
 * 
 * Command-line interface for encoding and decoding text files using the TCF format.
 */

import * as fs from 'fs';
import * as path from 'path';
import { TextCodec } from './tcf-codec';

function showUsage(): void {
  console.log(`
TCF Text Codec CLI

Usage:
  tcf encode <input.txt> <output.tcf>  - Encode text file to TCF format
  tcf decode <input.tcf> <output.txt>  - Decode TCF file to text
  tcf stats <input.txt>                - Show compression statistics
  tcf help                             - Show this help message

Examples:
  tcf encode document.txt document.tcf
  tcf decode document.tcf document_decoded.txt
  tcf stats document.txt
`);
}

function encodeFile(inputPath: string, outputPath: string): void {
  try {
    console.log(`Encoding ${inputPath} to ${outputPath}...`);
    
    const inputText = fs.readFileSync(inputPath, 'utf8');
    const tcfData = TextCodec.encode(inputText);
    
    fs.writeFileSync(outputPath, tcfData);
    
    const stats = TextCodec.getStats(inputText, tcfData);
    console.log(`✓ Encoding completed successfully!`);
    console.log(`  Original size: ${stats.originalSize} bytes`);
    console.log(`  Compressed size: ${stats.compressedSize} bytes`);
    console.log(`  Compression ratio: ${stats.compressionRatio.toFixed(2)}:1`);
    console.log(`  Space savings: ${stats.savings.toFixed(1)}%`);
    
  } catch (error: any) {
    console.error(`✗ Encoding failed: ${error.message}`);
    process.exit(1);
  }
}

function decodeFile(inputPath: string, outputPath: string): void {
  try {
    console.log(`Decoding ${inputPath} to ${outputPath}...`);
    
    const tcfData = fs.readFileSync(inputPath);
    const decodedText = TextCodec.decode(tcfData);
    
    fs.writeFileSync(outputPath, decodedText, 'utf8');
    
    console.log(`✓ Decoding completed successfully!`);
    console.log(`  Output size: ${Buffer.from(decodedText, 'utf8').length} bytes`);
    
  } catch (error: any) {
    console.error(`✗ Decoding failed: ${error.message}`);
    process.exit(1);
  }
}

function showStats(inputPath: string): void {
  try {
    console.log(`Analyzing ${inputPath}...`);
    
    const inputText = fs.readFileSync(inputPath, 'utf8');
    const tcfData = TextCodec.encode(inputText);
    const stats = TextCodec.getStats(inputText, tcfData);
    
    console.log(`\nCompression Statistics:`);
    console.log(`  Original size: ${stats.originalSize} bytes`);
    console.log(`  Compressed size: ${stats.compressedSize} bytes`);
    console.log(`  Compression ratio: ${stats.compressionRatio.toFixed(2)}:1`);
    console.log(`  Space savings: ${stats.savings.toFixed(1)}%`);
    
    // Character analysis
    const charCount = inputText.length;
    const uniqueChars = new Set(inputText).size;
    const entropy = calculateEntropy(inputText);
    
    console.log(`\nText Analysis:`);
    console.log(`  Character count: ${charCount}`);
    console.log(`  Unique characters: ${uniqueChars}`);
    console.log(`  Entropy: ${entropy.toFixed(2)} bits/char`);
    
  } catch (error: any) {
    console.error(`✗ Analysis failed: ${error.message}`);
    process.exit(1);
  }
}

function calculateEntropy(text: string): number {
  const frequencies = new Map<string, number>();
  
  // Count character frequencies
  for (const char of text) {
    frequencies.set(char, (frequencies.get(char) || 0) + 1);
  }
  
  // Calculate entropy
  let entropy = 0;
  const length = text.length;
  
  for (const frequency of frequencies.values()) {
    const probability = frequency / length;
    entropy -= probability * Math.log2(probability);
  }
  
  return entropy;
}

// Main CLI logic
function main(): void {
  const args = process.argv.slice(2);
  
  if (args.length === 0 || args[0] === 'help') {
    showUsage();
    return;
  }
  
  const command = args[0];
  
  switch (command) {
    case 'encode':
      if (args.length !== 3) {
        console.error('Error: encode command requires input and output file paths');
        showUsage();
        process.exit(1);
      }
      encodeFile(args[1], args[2]);
      break;
      
    case 'decode':
      if (args.length !== 3) {
        console.error('Error: decode command requires input and output file paths');
        showUsage();
        process.exit(1);
      }
      decodeFile(args[1], args[2]);
      break;
      
    case 'stats':
      if (args.length !== 2) {
        console.error('Error: stats command requires input file path');
        showUsage();
        process.exit(1);
      }
      showStats(args[1]);
      break;
      
    default:
      console.error(`Error: Unknown command '${command}'`);
      showUsage();
      process.exit(1);
  }
}

if (require.main === module) {
  main();
}

File: src/bin/tcf_cli.rs
use clap::{Arg, Command};
use codec_cdn_rust::codecs::text::TcfCodec;
use std::fs;
use std::io::{self, Read};
use std::path::Path;

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let matches = Command::new("tcf-cli")
        .version("1.0")
        .about("Text Codec Format (TCF) CLI tool with advanced arithmetic coding")
        .subcommand(
            Command::new("encode")
                .about("Encode text file to TCF format")
                .arg(
                    Arg::new("input")
                        .help("Input text file (or '-' for stdin)")
                        .required(true)
                        .value_name("FILE")
                )
                .arg(
                    Arg::new("output")
                        .help("Output TCF file")
                        .required(true)
                        .value_name("FILE")
                )
        )
        .subcommand(
            Command::new("decode")
                .about("Decode TCF file to text")
                .arg(
                    Arg::new("input")
                        .help("Input TCF file")
                        .required(true)
                        .value_name("FILE")
                )
                .arg(
                    Arg::new("output")
                        .help("Output text file (or '-' for stdout)")
                        .required(true)
                        .value_name("FILE")
                )
        )
        .subcommand(
            Command::new("info")
                .about("Show information about TCF file")
                .arg(
                    Arg::new("input")
                        .help("Input TCF file")
                        .required(true)
                        .value_name("FILE")
                )
        )
        .get_matches();

    match matches.subcommand() {
        Some(("encode", sub_matches)) => {
            let input = sub_matches.get_one::<String>("input").unwrap();
            let output = sub_matches.get_one::<String>("output").unwrap();
            
            let text = if input == "-" {
                let mut buffer = String::new();
                io::stdin().read_to_string(&mut buffer)?;
                buffer
            } else {
                fs::read_to_string(input)?
            };
            
            println!("Encoding {} characters...", text.len());
            
            let compressed = TcfCodec::encode(&text)?;
            fs::write(output, &compressed)?;
            
            let stats = TcfCodec::get_stats(&text, &compressed);
            println!("✓ Encoding complete!");
            println!("  Input: {} bytes", text.as_bytes().len());
            println!("  Output: {} bytes", compressed.len());
            println!("  Compression ratio: {:.2}:1", stats.compression_ratio);
            println!("  Space savings: {:.2}%", stats.savings_percent);
        }
        
        Some(("decode", sub_matches)) => {
            let input = sub_matches.get_one::<String>("input").unwrap();
            let output = sub_matches.get_one::<String>("output").unwrap();
            
            let compressed = fs::read(input)?;
            println!("Decoding {} bytes...", compressed.len());
            
            let text = TcfCodec::decode(&compressed)?;
            
            if output == "-" {
                print!("{}", text);
            } else {
                fs::write(output, &text)?;
            }
            
            println!("✓ Decoding complete!");
            println!("  Decoded {} characters", text.len());
        }
        
        Some(("info", sub_matches)) => {
            let input = sub_matches.get_one::<String>("input").unwrap();
            
            let compressed = fs::read(input)?;
            let header = TcfCodec::parse_header(&compressed)?;
            
            println!("TCF File Information:");
            println!("  Magic: {}", header.magic);
            println!("  Version: {}", header.version);
            println!("  Original size: {} bytes", header.original_size);
            println!("  Compressed size: {} bytes", header.compressed_size);
            println!("  Compression method: {}", header.compression_method);
            println!("  Model size: {} bytes", header.model_size);
            println!("  Checksum: {}", header.checksum);
            
            let compression_ratio = header.original_size as f64 / compressed.len() as f64;
            let savings = ((header.original_size - compressed.len() as u64) as f64 / header.original_size as f64) * 100.0;
            println!("  Compression ratio: {:.2}:1", compression_ratio);
            println!("  Space savings: {:.2}%", savings);
        }
        
        _ => {
            eprintln!("No subcommand provided. Use --help for usage information.");
            std::process::exit(1);
        }
    }
    
    Ok(())
}

// Usage examples:
// echo "Hello, World!" | tcf-cli encode - hello.tcf
// tcf-cli decode hello.tcf -
// tcf-cli info hello.tcf